### PR TITLE
Added classes for Sync and Async infer requests

### PR DESCRIPTION
**Pull request by ilyachur**
### Details:
 - *item1*
 - *...*

### Tickets:
 - *ticket-id*

### CODE REVIEW

File: src/inference/dev_api/openvino/runtime/iasync_infer_request.hpp [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+ * @brief OpenVINO Runtime AsyncInferRequest interface
+ * @file openvino/runtime/iasync_nfer_request.hpp
+ */
+
+#pragma once
+
+#include <future>
+#include <memory>
+
+#include "ie_common.h"
```
Comment (ilya-lavrenov) [inline src/inference/dev_api/openvino/runtime/iasync_infer_request.hpp:15]: do we need old include here?

File: src/inference/dev_api/openvino/runtime/iasync_infer_request.hpp [hunk 2]
```diff
@@ -0,0 +1,276 @@
+// Copyright (C) 2018-2022 Intel Corporation
```
Comment (ilya-lavrenov) [inline src/inference/dev_api/openvino/runtime/iasync_infer_request.hpp:1]: years..

File: src/inference/dev_api/openvino/runtime/iinfer_request.hpp [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+/**
+ * @brief OpenVINO Runtime InferRequest interface
+ * @file openvino/runtime/iinfer_request.hpp
+ */
+
+#pragma once
+
+#include <exception>
+#include <memory>
+#include <openvino/runtime/tensor.hpp>
```
Comment (ilya-lavrenov) [inline src/inference/dev_api/openvino/runtime/iinfer_request.hpp:14]: use ""

File: src/inference/src/dev/converter_utils.cpp [hunk 4]
```diff
@@ -397,3 +411,296 @@ std::shared_ptr<ov::ICompiledModel> ov::legacy_convert::convert_compiled_model(
...
+    ov::Output<const ov::Node> find_port(const std::string& legacy_name) const {
+        for (const auto& port : m_request->get_inputs()) {
+            if (get_legacy_name_from_port(port) == legacy_name)
+                return port;
+        }
+        for (const auto& port : m_request->get_outputs()) {
+            if (get_legacy_name_from_port(port) == legacy_name)
+                return port;
+        }
+        // TODO:
```
Comment (ilya-lavrenov) [inline src/inference/src/dev/converter_utils.cpp:427]: should we fix todo?

File: src/inference/src/dev/converter_utils.cpp [hunk 5]
```diff
@@ -397,3 +411,296 @@ std::shared_ptr<ov::ICompiledModel> ov::legacy_convert::convert_compiled_model(
...
+        m_request->infer();
+    }
+
+    void Cancel() override {
+        m_request->cancel();
+    }
+
+    std::map<std::string, InferenceEngine::InferenceEngineProfileInfo> GetPerformanceCounts() const override {
+        OPENVINO_NOT_IMPLEMENTED;
+        // TODO:
```
Comment (ilya-lavrenov) [inline src/inference/src/dev/converter_utils.cpp:445]: should we convert new structure to old one?

File: src/inference/src/dev/converter_utils.cpp [hunk 6]
```diff
@@ -397,3 +411,296 @@ std::shared_ptr<ov::ICompiledModel> ov::legacy_convert::convert_compiled_model(
...
+    ov::Tensor get_tensor(const ov::Output<const ov::Node>& port) const override {
+        const auto& name = get_legacy_name_from_port(port);
+        OPENVINO_ASSERT(!m_request->GetBlobs(name),
+                        "get_tensor shall not be used together with batched "
+                        "set_tensors/set_input_tensors for name '",
+                        name,
+                        "'");
+        auto blob = m_request->GetBlob(name);
+        // soVec = {_so, _impl->getPointerToSo()};
+        // Tensor tensor = {blob, soVec};
```
Comment (ilya-lavrenov) [inline src/inference/src/dev/converter_utils.cpp:625]: remove commented code?

File: src/inference/src/dev/converter_utils.cpp [hunk 7]
```diff
@@ -397,3 +411,296 @@ std::shared_ptr<ov::ICompiledModel> ov::legacy_convert::convert_compiled_model(
...
+        m_request->SetBlob(get_legacy_name_from_port(port), tensor._impl);
+    }
+
+    std::vector<ov::Tensor> get_tensors(const ov::Output<const ov::Node>& port) const override {
+        auto blobs = m_request->GetBlobs(get_legacy_name_from_port(port));
+        std::vector<ov::Tensor> ret;
+        if (!blobs)
+            return ret;
+        for (size_t i = 0; i < blobs->size(); i++) {
+            ret.emplace_back(ov::Tensor{blobs->getBlob(i), {}});
```
Comment (ilya-lavrenov) [inline src/inference/src/dev/converter_utils.cpp:639]: should not we also add getPointerToSo() here in some other places in this class?

File: src/inference/src/dev/converter_utils.hpp [hunk 8]
```diff
@@ -22,11 +24,16 @@ std::shared_ptr<const ov::Model> convert_model(const InferenceEngine::CNNNetwork
...
-std::shared_ptr<::InferenceEngine::IExecutableNetworkInternal> convert_compiled_model(
+OPENVINO_RUNTIME_API std::shared_ptr<::InferenceEngine::IExecutableNetworkInternal> convert_compiled_model(
     const std::shared_ptr<::ov::ICompiledModel>& model);
 std::shared_ptr<::ov::ICompiledModel> convert_compiled_model(
     const std::shared_ptr<::InferenceEngine::IExecutableNetworkInternal>& model);
 
+OPENVINO_RUNTIME_API std::shared_ptr<::InferenceEngine::IInferRequestInternal> convert_infer_request(
+    const std::shared_ptr<::ov::IAsyncInferRequest>& request);
+OPENVINO_RUNTIME_API std::shared_ptr<::ov::IAsyncInferRequest> convert_infer_request(
+    const std::shared_ptr<::InferenceEngine::IInferRequestInternal>& request);
```
Comment (ilya-lavrenov) [inline src/inference/src/dev/converter_utils.hpp:35]: do we use these methods outside of inference library?

File: src/plugins/template/src/CMakeLists.txt [hunk 9]
```diff
@@ -26,6 +26,7 @@ ov_mark_target_as_cc(${TARGET_NAME})
 
 target_include_directories(${TARGET_NAME} PRIVATE
     "${CMAKE_CURRENT_SOURCE_DIR}"
+    "${OpenVINO_SOURCE_DIR}/src/inference/src/dev"
```
Comment (ilya-lavrenov) [inline src/plugins/template/src/CMakeLists.txt:29]: why is it needed? temp solution until template plugin migrate to new Infer Request?

File: src/plugins/template/src/compiled_model.cpp [hunk 10]
```diff
@@ -124,17 +128,8 @@ void TemplatePlugin::CompiledModel::compile_model(const std::shared_ptr<ov::Mode
...
-        std::static_pointer_cast<TemplatePlugin::TemplateInferRequest>(internal_request),
-        get_task_executor(),
-        get_template_plugin()->_waitExecutor,
-        get_callback_executor());
-}
-
-std::shared_ptr<InferenceEngine::IInferRequestInternal> TemplatePlugin::CompiledModel::create_sync_infer_request()
-    const {
+std::shared_ptr<ov::IAsyncInferRequest> TemplatePlugin::CompiledModel::create_infer_request() const {
+    // auto internal_request = create_sync_infer_request();
```
Comment (ilya-lavrenov) [inline src/plugins/template/src/compiled_model.cpp:132]: remove commented code?

File: src/inference/src/dev/converter_utils.hpp (hunk 8 above)
Comment (ilyachur) [reply src/inference/src/dev/converter_utils.hpp:35]: Yes it is temporary change for template plugin.

File: src/plugins/template/src/CMakeLists.txt (hunk 9 above)
Comment (ilyachur) [reply src/plugins/template/src/CMakeLists.txt:29]: Yes it is temporary solution

File: src/plugins/template/src/compiled_model.cpp (hunk 10 above)
Comment (ilyachur) [reply src/plugins/template/src/compiled_model.cpp:132]: This code will used after the migration to new API

**Review by ilya-lavrenov — APPROVED** [review]

**Pull request merged**